[bgfx] Update to 1.121.8534-453

bwrsandman opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [x] The "supports" clause reflects platforms that may be fixed by this new version
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

Add two helpful warnings about bad use of features:
1. Build tools on host when cross compiling: tools take a while to build and they're there to be run on the host during build time (e.g. shader compiler and binary generators)
2. Don't use the multithread feature on webassembly because there are no `bx::Thread` definitions on that platform.

Also added a cross compile check and definition to allow the bgfx cmake to generate the proper config file for shader header includes.

bwrsandman committed:
[bgfx] Update to v1.121.8534-452

dg0yt commented on ports/bgfx/portfile.cmake:
```diff
@@ -44,8 +44,21 @@ else ()
   set(BGFX_LIBRARY_TYPE "STATIC")
 endif ()
 
+if (BGFX_CONFIG_MULTITHREADED AND TARGET_TRIPLET MATCHES "(emscripten)")
```
Can't this be handled in the manifest?
Otherwise don't match the triplet name: Use `VCPKG_TARGET_IS_EMSCRIPTEN`.

bwrsandman replied:
How would that be done in the manifest?

It should be on by default on all platforms but always off for emscripten.
The reason why it's an option is that it is much easier to debug when things go wrong if multithreading is off.

dg0yt commented on ports/bgfx/portfile.cmake:
```diff
@@ -44,8 +44,21 @@ else ()
+
 file(COPY "${CMAKE_CURRENT_LIST_DIR}/vcpkg-inject-packages.cmake" DESTINATION "${SOURCE_PATH}")
 
+if("${_HOST_TRIPLET}" STREQUAL "${TARGET_TRIPLET}")
```
This the portfile: we have VCPKG_CROSSCOMPILING`.

dg0yt commented on ports/bgfx/portfile.cmake:
```diff
@@ -44,8 +44,21 @@ else ()
+if("${_HOST_TRIPLET}" STREQUAL "${TARGET_TRIPLET}")
+  set(CMAKE_CROSSCOMPILING OFF)
+else()
+if (BGFX_BUILD_TOOLS)
+  message(WARNING "If cross compiling (even x86 on x64), install use the tools (bin2c, shaderc) from the host triplet")
+endif ()
```
Shouldn't this get some indent?

dg0yt commented on ports/bgfx/portfile.cmake:
```diff
@@ -56,6 +69,7 @@ vcpkg_cmake_configure(
           -DBGFX_AMALGAMATED=ON
           -DBGFX_BUILD_EXAMPLES=OFF
           -DBGFX_OPENGLES_VERSION=30
+          -DCMAKE_CROSSCOMPILING=${CMAKE_CROSSCOMPILING}
```
I would prefer to leave `CMAKE_CROSSCOMPILING` to toolchain initialization.

bwrsandman replied:
The bgfx.cmake cmake files don't get `CMAKE_CROSSCOMPILING` set when built from vcpkg for whatever reason.
This is the solution I came up with without specifically putting vcpkg-specific syntax in it.
Could also do a patch but that seems unnecessary if I can just turn on this define.

BillyONeal replied:
I believe it is intentional that these don't always match. `VCPKG_CROSSCOMPILING` means that the host triplet and target triplet are different, but `CMAKE_CROSSCOMPILING` means that programs built for this target can't be run *at all* on the host system. Consider target == x86-windows and host == x64-windows. That is `VCPKG_CROSSCOMPILING` because host and target differ, but we might want to not set `CMAKE_CROSSCOMPILING` as lots of projects get confused when that is set targeting Windows.

This is a toolchain decision. What configuration were you testing where that was not set?

bwrsandman replied:
Since bgfx tools makes 3 or 4 different binaries that we supply as "compilers" (including a shader compiler) in the cmake configuration.
This allows projects to call via cmake script macro those compilers to compile shaders, textures, meshes to headers which bgfx can then load internally as string data.
We are able to offload a lot of processing to compile time and avoid issues of "asset" folders shipped with binaries.

There's an issue when cross compiling (x86 on x64 or arm or x64 or android/emscripten etc on x64-linux/x64-windows).
We want the host version of these compilers to be used to generate these headers.

bgfx.cmake relies on `CMAKE_CROSSCOMPILING` to determine whether or not to include the compilers. We would not want the compilers to be enabled in cross compilation.

As for x86 and x64. I don't mind it being treated as cross compiling and requiring x64 tools.

Hope that all makes sense.

BillyONeal replied:
> Your host machine might not be able to run those tools if they're cross compiled.

Yes, it might not be able to. But it might be able to too. What you're describing is what `CMAKE_CROSSCOMPILING` is for. But it's a toolchain decision, not a triplet decision.

bwrsandman replied:
I get your point.
Technically you could run arm on x64 linux tools if you have qemu.
Unless there's an easy way to determine all that using cmake, I'd rather only use the tools on the native arch.

dg0yt replied:
Hm, we don't need qemu to run x64-linux-dynamic on x64-linux. But this is `VCPKG_CROSSCOMPILING`, and in general we do use host triplet tools in this case.
I just feel that this should have a more explicit control than `CMAKE_CROSSCOMPILING`, at least in future revisions.

Cheney-W commented:
x64-osx error:
```
/Users/vagrant/Data/buildtrees/bgfx/src/3db6bda3f6-a485e42544.clean/src/renderer_mtl.mm:560:37: error: use of undeclared identifier 'MTLGPUFamilyApple8'
                                                if ([m_device supportsFamily: MTLGPUFamilyApple8])
                                                                              ^
/Users/vagrant/Data/buildtrees/bgfx/src/3db6bda3f6-a485e42544.clean/src/renderer_mtl.mm:2340:40: error: property 'vertexBindings' not found on object of type 'MTLRenderPipelineReflection *'
                                                processArguments(pso, reflection.vertexBindings, reflection.fragmentBindings);
                                                                                 ^
/Users/vagrant/Data/buildtrees/bgfx/src/3db6bda3f6-a485e42544.clean/src/renderer_mtl.mm:2340:67: error: property 'fragmentBindings' not found on object of type 'MTLRenderPipelineReflection *'
                                                processArguments(pso, reflection.vertexBindings, reflection.fragmentBindings);
                                                                                                            ^
/Users/vagrant/Data/buildtrees/bgfx/src/3db6bda3f6-a485e42544.clean/src/renderer_mtl.mm:2387:38: error: property 'bindings' not found on object of type 'MTLComputePipelineReflection *'
                                processArguments(pso, reflection.bindings, NULL);
```
Error logs: https://dev.azure.com/vcpkg/public/_build/results?buildId=90944&view=artifacts&pathAsName=false&type=publishedArtifacts

Cheney-W commented:
Moving the PR to draft. Please set it to "ready for review" once changes have been applied.

bwrsandman committed:
[bgfx] Use manifest to prevent features on cross compile

bwrsandman commented:
@Cheney-W the osx build machine has MacOSX12.1.sdk perhaps, it is too old to have a definition for `MTLGPUFamilyApple8`?

Cheney-W commented:
@vicroms Do we have plans to upgrade the osx machine?

vicroms commented on ports/bgfx/vcpkg.json:
```diff
@@ -24,11 +24,13 @@
   ],
   "features": {
     "multithreaded": {
-      "description": "Encode and render on different threads"
+      "description": "Encode and render on different threads",
+      "supports": "!emscripten"
     },
```
```no-highlight
"default-features": [
  "platform-default-features"
],
"features": {
  "platform-default-features": {
    "description": "Enables default features per platform",
    "dependencies": [{
      "name": "bgfx",
      "features": ["multithreaded"],
      "platform": "!empscripten"
    }]
  },
  "multithreaded": {
    "description": "Encode and render on different threads",
    "supports": "!empscripten"
  }
}
```

Do something like this instead to disable the feature on emscripten.

vicroms commented on ports/bgfx/portfile.cmake:
```diff
@@ -44,6 +44,14 @@ else ()
   set(BGFX_LIBRARY_TYPE "STATIC")
 endif ()
 
+if (BGFX_CONFIG_MULTITHREADED AND VCPKG_TARGET_IS_EMSCRIPTEN)
+    message(WARNING "bgfx does not support multithreading. Unset the multithreaded default feature.")
+endif ()
+
+if (VCPKG_CROSSCOMPILING AND BGFX_BUILD_TOOLS)
+    message(WARNING "If cross compiling (even x86 on x64), install use the tools (bin2c, shaderc) from the host triplet")
+endif ()
```
These warnings are not necessary since the support expressions on vcpkg.json will prevent them.

Cheney-W commented:
@vicroms If we currently don't have any plans to modify the OSX machine environment, can we add `bgfx:x64-osx=fail` to ci.baseline.txt?

vicroms commented:
Please add to baseline. Billy has been in talks with the infrastructure team to get us better macOS support but doesn't look like it's going to get done soon.

Cheney-W commented:
Moving the PR to draft. Please set it to "ready for review" once changes have been applied.

bwrsandman committed:
[bgfx] Update version

bwrsandman committed:
[bgfx] Blacklist osx test because of missing MTLGPUFamilyApple8

Cheney-W commented:
Waiting for the CI baseline issue of halide:x86-windows is fixed.

bwrsandman commented:
Can we move forward with this PR?

vicroms commented:
/azp run

bwrsandman committed:
[bgfx] Windows platforms must compile static

Even if tools aren't build.

bwrsandman commented:
- Removed the need for `CMAKE_CROSSCOMPILING`, thanks for the review and discussion.
- Fixed a bug with windows linking for c++ - bgfx doesn't export the c++ functions and therefore cannot be linked statically for dlls
- Bump up version to 1.121.8534-453 which has fixes for bgfx::bx headers installation and how tools are located in cross compile context

bwrsandman committed:
[bgfx]: Add bx fix for amalgamated build

BillyONeal commented:
Thanks for the update!
